backend/stakepoold/rpc/rpcserver/client.go

```diff
@@ -0,0 +1,127 @@
+package rpcserver
+
+import (
+	"sync"
+	"time"
+
+	"github.com/decred/dcrd/rpcclient/v3"
+)
+
+const (
+	// connectionRetryInterval is the amount of time to wait in between
+	// retries when automatically reconnecting to an RPC server.
+	connectionRetryInterval = time.Second * 5
+	// disconnectCheckInterval is the amount of time to wait between
+	// checks for a disconnection.
+	disconnectCheckInterval = time.Second * 10
+)
+
+type Client struct {
+	*rpcclient.Client
+	cfg          *rpcclient.ConnConfig
+	ntfnHandlers *rpcclient.NotificationHandlers
+	wg           sync.WaitGroup
+	stop         chan struct{}
+	disconnected chan struct{}
+}
+
+// New creates a new Client and starts the automatic reconnection handler.
+// Returns an error if unable to construct a new rpcclient.
+func New(cfg *rpcclient.ConnConfig, ntfnHandlers *rpcclient.NotificationHandlers) (*Client, error) {
+	client, err := rpcclient.New(cfg, ntfnHandlers)
+	if err != nil {
+		return nil, err
+	}
+	c := &Client{
+		Client:       client,
+		cfg:          cfg,
+		ntfnHandlers: ntfnHandlers,
+		stop:         make(chan struct{}),
+		disconnected: make(chan struct{}),
+	}
+	go func(c *Client) {
+		c.checkForDisconnect()
+	}(c)
+	go func(c *Client) {
+		c.wg.Add(1)
+		c.autoReconnect()
+	}(c)
+	return c, nil
+}
+
+// Stop stops automatic reconnections.
+func (c *Client) Stop() {
+	c.stop <- struct{}{}
+	c.wg.Wait()
+}
+
+// checkForDisconnect checks for a client disconnect every connectionRetryInterval.
+// If disconnected sends a value to the disconnect channel if not currently
+// trying to reconnect.
+//
+// This function must be run as a goroutine.
+func (c *Client) checkForDisconnect() {
+	for {
+		time.Sleep(disconnectCheckInterval)
+		if c.Disconnected() {
+			select {
+			case c.disconnected <- struct{}{}:
+			default:
+			}
+		}
+	}
+}
+
+// autoReconnect waits for a disconnect or stop. On disconnect it attempts to
+// reconnect to the client every connectionRetryInterval.
+//
+// This function must be run as a goroutine.
+func (c *Client) autoReconnect() {
+out:
+	for {
+		select {
+		case <-c.stop:
+			break out
+		case <-c.disconnected:
+			// Client is disconnected. Try to reconnect.
+		}
+
+	reconnect:
+		for {
+			select {
+			case <-c.stop:
+				break out
+			default:
+			}
+
+			// Properly shutdown old client.
+			c.Shutdown()
+			c.WaitForShutdown()
+
+			// Start a new client.
+			client, err := rpcclient.New(c.cfg, c.ntfnHandlers)
+
+			if err != nil {
+				log.Infof("Failed to connect to %s: %v",
+					c.cfg.Host, err)
+
+				log.Infof("Retrying connection to %s in "+
+					"%s", c.cfg.Host, connectionRetryInterval)
+				time.Sleep(connectionRetryInterval)
+				continue reconnect
+			}
+
+			// Switch the new client with the old, shutdown one.
+			c.Client = client
```

Comment:
Do we need a mutex on c.Client to make this safe for concurrent access?

Reply:
probably? And then this should be changed to copy the rpcclient value?  I don't know what would happen if we took the address before a change and used it after. Panic? I was thinking to use a rwmutex here, use RLock for all commands and lock here, to let the client deal with the concurrency related to commands.

Reply:
or does passing an address keep the old client alive until everyone is done with it? In which case maybe this is ok. I'll make some tests to see what happens. This is basic stuff I should know anyway.

Reply:
I think it's ok as an address. https://play.golang.org/p/110tx7abKJa
Does the address need a lock?

Reply:
I've not reviewed this PR, but the linked play is a race.

Reply:
> or does passing an address keep the old client alive until everyone is done with it? In which case maybe this is ok. I'll make some tests to see what happens. This is basic stuff I should know anyway.

As long as the old client is referenced by a pointer somewhere then it a valid object.  But that doesn't mean concurrent read/write of `c.Client` is OK.  Here you write to it, but in another goroutine there is access to `c.Client.  You do not want the possibility of the write occurring at the same time as the read (or another write for that matter)!